ddtrace/tracer/rules_sampler.go

```diff
@@ -0,0 +1,132 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2016 Datadog, Inc.
+
+package tracer
+
+import (
+	"math"
+	"os"
+	"strconv"
+	"time"
+
+	"golang.org/x/time/rate"
+
+	"gopkg.in/DataDog/dd-trace-go.v1/ddtrace/ext"
+	"gopkg.in/DataDog/dd-trace-go.v1/internal/log"
+	"gopkg.in/DataDog/dd-trace-go.v1/internal/samplernames"
+)
+
+// traceRulesSampler allows a user-defined list of rules to apply to spans.
+// These rules can match based on the span's Service, Name or both.
+// When making a sampling decision, the rules are checked in order until
+// a match is found.
+// If a match is found, the rate from that rule is used.
+// If no match is found, and the DD_TRACE_SAMPLE_RATE environment variable
+// was set to a valid rate, that value is used.
+// Otherwise, the rules sampler didn't apply to the span, and the decision
+// is passed to the priority sampler.
+//
+// The rate is used to determine if the span should be sampled, but an upper
+// limit can be defined using the DD_TRACE_RATE_LIMIT environment variable.
+// Its value is the number of spans to sample per second.
+// Spans that matched the rules but exceeded the rate limit are not sampled.
+type traceRulesSampler struct {
+	rules      []SamplingRule // the rules to match spans with
+	globalRate float64        // a rate to apply when no rules match a span
+	limiter    *rateLimiter   // used to limit the volume of spans sampled
+}
+
+// newTraceRulesSampler configures a *traceRulesSampler instance using the given set of rules.
+// Invalid rules or environment variable values are tolerated, by logging warnings and then ignoring them.
+func newTraceRulesSampler(rules []SamplingRule) *traceRulesSampler {
+	return &traceRulesSampler{
+		rules:      rules,
+		globalRate: globalSampleRate(),
+		limiter:    newRateLimiter(),
+	}
+}
+
+func (rs *traceRulesSampler) enabled() bool {
+	return len(rs.rules) > 0 || !math.IsNaN(rs.globalRate)
+}
+
+// apply uses the sampling rules to determine the sampling rate for the
+// provided span. If the rules don't match, and a default rate hasn't been
+// set using DD_TRACE_SAMPLE_RATE, then it returns false and the span is not
+// modified.
+func (rs *traceRulesSampler) apply(span *span) bool {
+	if !rs.enabled() {
+		// short path when disabled
+		return false
+	}
+
+	var matched bool
+	rate := rs.globalRate
+	for _, rule := range rs.rules {
+		if rule.match(span) {
```

Comment:
Gotcha, okay. I didn't realize that when I was reviewing it. In that case, can you file an issue with those details and add a TODO to look into this later. I do suspect there could be a bug here, so it would be good to follow up.